// viem/_esm/zksync/index.js
// biome-ignore lint/performance/noBarrelFile: entrypoint module
export { toSmartAccount, } from './accounts/toSmartAccount.js';
export { toMultisigSmartAccount, } from './accounts/toMultisigSmartAccount.js';
export { toSinglesigSmartAccount, } from './accounts/toSinglesigSmartAccount.js';
export { deployContract, } from './actions/deployContract.js';
export { encodeDeployData, } from './utils/abi/encodeDeployData.js';
export { hashBytecode, } from './utils/hashBytecode.js';
export { estimateFee, } from './actions/estimateFee.js';
export { getAllBalances, } from './actions/getAllBalances.js';
export { getBlockDetails, } from './actions/getBlockDetails.js';
export { getDefaultBridgeAddresses, } from './actions/getDefaultBridgeAddresses.js';
export { getBridgehubContractAddress } from './actions/getBridgehubContractAddress.js';
export { getL1Allowance, } from './actions/getL1Allowance.js';
export { getL1Balance, } from './actions/getL1Balance.js';
export { getL1BatchBlockRange, } from './actions/getL1BatchBlockRange.js';
export { getL1BatchDetails, } from './actions/getL1BatchDetails.js';
export { getL1BatchNumber } from './actions/getL1BatchNumber.js';
export { getL1ChainId } from './actions/getL1ChainId.js';
export { getL1TokenBalance, } from './actions/getL1TokenBalance.js';
export { getLogProof, } from './actions/getLogProof.js';
export { getMainContractAddress } from './actions/getMainContractAddress.js';
export { getRawBlockTransactions, } from './actions/getRawBlockTransactions.js';
export { getTestnetPaymasterAddress } from './actions/getTestnetPaymasterAddress.js';
export { getTransactionDetails, } from './actions/getTransactionDetails.js';
export { requestExecute, } from './actions/requestExecute.js';
export { sendTransaction, } from './actions/sendTransaction.js';
export { sendEip712Transaction, } from './actions/sendEip712Transaction.js';
export { signEip712Transaction, } from './actions/signEip712Transaction.js';
export { signTransaction, } from './actions/signTransaction.js';
export { getL2TokenAddress, } from './actions/getL2TokenAddress.js';
export { getL1TokenAddress, } from './actions/getL1TokenAddress.js';
export { withdraw, } from './actions/withdraw.js';
export { finalizeWithdrawal, } from './actions/finalizeWithdrawal.js';
export { legacyEthAddress, l2BaseTokenAddress, } from './constants/address.js';
export { 
/** @deprecated Use `zksync` instead */
zksync as zkSync, zksync, 
/** @deprecated Use `zksync` instead */
zksyncInMemoryNode as zkSyncInMemoryNode, zksyncLocalCustomHyperchain, zksyncLocalHyperchain, zksyncLocalHyperchainL1, zksyncInMemoryNode, 
/** @deprecated Use `zksync` instead */
zksyncLocalNode as zkSyncLocalNode, zksyncLocalNode, 
/** @deprecated Use `zksync` instead */
zksyncSepoliaTestnet as zkSyncSepoliaTestnet, zksyncSepoliaTestnet, } from './chains.js';
export { chainConfig } from './chainConfig.js';
export { eip712WalletActions, } from './decorators/eip712.js';
export { publicActionsL1, } from './decorators/publicL1.js';
export { publicActionsL2, } from './decorators/publicL2.js';
export { walletActionsL1, } from './decorators/walletL1.js';
export { walletActionsL2, } from './decorators/walletL2.js';
export { serializeTransaction } from './serializers.js';
export { getL2HashFromPriorityOp, } from './utils/bridge/getL2HashFromPriorityOp.js';
export { getApprovalBasedPaymasterInput, } from './utils/paymaster/getApprovalBasedPaymasterInput.js';
export { getGeneralPaymasterInput, } from './utils/paymaster/getGeneralPaymasterInput.js';
export { parseEip712Transaction } from './utils/parseEip712Transaction.js';
