// @polkadot/util/cjs/bn/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bnToU8a = exports.bnToHex = exports.bnToBn = exports.bnSqrt = exports.bnMin = exports.bnMax = exports.bnFromHex = exports.BN = void 0;
const tslib_1 = require("tslib");
/**
 * @summary Utility methods to convert to and from `BN` objects
 */
var bn_js_1 = require("./bn.js");
Object.defineProperty(exports, "BN", { enumerable: true, get: function () { return bn_js_1.BN; } });
var fromHex_js_1 = require("./fromHex.js");
Object.defineProperty(exports, "bnFromHex", { enumerable: true, get: function () { return fromHex_js_1.bnFromHex; } });
var min_js_1 = require("./min.js");
Object.defineProperty(exports, "bnMax", { enumerable: true, get: function () { return min_js_1.bnMax; } });
Object.defineProperty(exports, "bnMin", { enumerable: true, get: function () { return min_js_1.bnMin; } });
var sqrt_js_1 = require("./sqrt.js");
Object.defineProperty(exports, "bnSqrt", { enumerable: true, get: function () { return sqrt_js_1.bnSqrt; } });
var toBn_js_1 = require("./toBn.js");
Object.defineProperty(exports, "bnToBn", { enumerable: true, get: function () { return toBn_js_1.bnToBn; } });
var toHex_js_1 = require("./toHex.js");
Object.defineProperty(exports, "bnToHex", { enumerable: true, get: function () { return toHex_js_1.bnToHex; } });
var toU8a_js_1 = require("./toU8a.js");
Object.defineProperty(exports, "bnToU8a", { enumerable: true, get: function () { return toU8a_js_1.bnToU8a; } });
tslib_1.__exportStar(require("./consts.js"), exports);
