// @polkadot/util/cjs/object/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.objectValues = exports.objectSpread = exports.objectProperty = exports.objectProperties = exports.objectKeys = exports.objectEntries = exports.objectCopy = exports.objectClear = void 0;
var clear_js_1 = require("./clear.js");
Object.defineProperty(exports, "objectClear", { enumerable: true, get: function () { return clear_js_1.objectClear; } });
var copy_js_1 = require("./copy.js");
Object.defineProperty(exports, "objectCopy", { enumerable: true, get: function () { return copy_js_1.objectCopy; } });
var entries_js_1 = require("./entries.js");
Object.defineProperty(exports, "objectEntries", { enumerable: true, get: function () { return entries_js_1.objectEntries; } });
var keys_js_1 = require("./keys.js");
Object.defineProperty(exports, "objectKeys", { enumerable: true, get: function () { return keys_js_1.objectKeys; } });
var property_js_1 = require("./property.js");
Object.defineProperty(exports, "objectProperties", { enumerable: true, get: function () { return property_js_1.objectProperties; } });
Object.defineProperty(exports, "objectProperty", { enumerable: true, get: function () { return property_js_1.objectProperty; } });
var spread_js_1 = require("./spread.js");
Object.defineProperty(exports, "objectSpread", { enumerable: true, get: function () { return spread_js_1.objectSpread; } });
var values_js_1 = require("./values.js");
Object.defineProperty(exports, "objectValues", { enumerable: true, get: function () { return values_js_1.objectValues; } });
