// @polkadot-labs/hdkd/dist/index.js
// src/ecdsaCreateDerive.ts
import { createDerive, ecdsa, ecdsaDerive } from "@polkadot-labs/hdkd-helpers";
var ecdsaCreateDerive = (seed) => createDerive({
  seed,
  curve: ecdsa,
  derive: ecdsaDerive
});

// src/ed25519CreateDerive.ts
import {
  createDerive as createDerive2,
  ed25519,
  ed25519Derive
} from "@polkadot-labs/hdkd-helpers";
var ed25519CreateDerive = (seed) => createDerive2({
  seed,
  curve: ed25519,
  derive: ed25519Derive
});

// src/sr25519CreateDerive.ts
import {
  createDerive as createDerive3,
  sr25519,
  sr25519Derive
} from "@polkadot-labs/hdkd-helpers";
var sr25519CreateDerive = (seed) => createDerive3({
  seed,
  curve: sr25519,
  derive: sr25519Derive
});

// src/withNetworkAccount.ts
import {
  accountId,
  ss58Address,
  ss58PublicKey
} from "@polkadot-labs/hdkd-helpers";
var withNetworkAccount = (keyPair, prefix = 42) => ({
  ...keyPair,
  accountId: accountId(keyPair.publicKey),
  ss58Address: ss58Address(keyPair.publicKey, prefix),
  ss58PublicKey: ss58PublicKey(keyPair.publicKey, prefix)
});
export {
  ecdsaCreateDerive,
  ed25519CreateDerive,
  sr25519CreateDerive,
  withNetworkAccount
};
