// @polkadot-labs/hdkd/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ecdsaCreateDerive: () => ecdsaCreateDerive,
  ed25519CreateDerive: () => ed25519CreateDerive,
  sr25519CreateDerive: () => sr25519CreateDerive,
  withNetworkAccount: () => withNetworkAccount
});
module.exports = __toCommonJS(index_exports);

// src/ecdsaCreateDerive.ts
var import_hdkd_helpers = require("@polkadot-labs/hdkd-helpers");
var ecdsaCreateDerive = (seed) => (0, import_hdkd_helpers.createDerive)({
  seed,
  curve: import_hdkd_helpers.ecdsa,
  derive: import_hdkd_helpers.ecdsaDerive
});

// src/ed25519CreateDerive.ts
var import_hdkd_helpers2 = require("@polkadot-labs/hdkd-helpers");
var ed25519CreateDerive = (seed) => (0, import_hdkd_helpers2.createDerive)({
  seed,
  curve: import_hdkd_helpers2.ed25519,
  derive: import_hdkd_helpers2.ed25519Derive
});

// src/sr25519CreateDerive.ts
var import_hdkd_helpers3 = require("@polkadot-labs/hdkd-helpers");
var sr25519CreateDerive = (seed) => (0, import_hdkd_helpers3.createDerive)({
  seed,
  curve: import_hdkd_helpers3.sr25519,
  derive: import_hdkd_helpers3.sr25519Derive
});

// src/withNetworkAccount.ts
var import_hdkd_helpers4 = require("@polkadot-labs/hdkd-helpers");
var withNetworkAccount = (keyPair, prefix = 42) => ({
  ...keyPair,
  accountId: (0, import_hdkd_helpers4.accountId)(keyPair.publicKey),
  ss58Address: (0, import_hdkd_helpers4.ss58Address)(keyPair.publicKey, prefix),
  ss58PublicKey: (0, import_hdkd_helpers4.ss58PublicKey)(keyPair.publicKey, prefix)
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ecdsaCreateDerive,
  ed25519CreateDerive,
  sr25519CreateDerive,
  withNetworkAccount
});
