// @polkadot-api/polkadot-sdk-compat/dist/esm/index.mjs
import { parsed } from './parsed.mjs';
export { fromParsed, toParsed } from './parsed.mjs';
import { fixChainSpec } from './parsed-enhancers/chain-spec.mjs';
import { fixPrematureBlocks } from './parsed-enhancers/fix-premature-blocks.mjs';
import { fixDescendantValues } from './parsed-enhancers/fix-descendant-values.mjs';
import { fixUnorderedBlocks } from './parsed-enhancers/fix-unordered-blocks.mjs';
import { fixUnorderedEvents } from './parsed-enhancers/fix-unordered-events.mjs';
import { patchChainHeadEvents } from './parsed-enhancers/patch-chainhead-events.mjs';
import { unpinHash } from './parsed-enhancers/unpin-hash.mjs';
import { translate } from './parsed-enhancers/translate.mjs';
import { withNumericIds } from './parsed-enhancers/numeric-ids.mjs';
import * as methods from './methods.mjs';
export { methods };

const withPolkadotSdkCompat = parsed(
  withNumericIds,
  translate,
  fixUnorderedEvents,
  unpinHash,
  patchChainHeadEvents,
  fixPrematureBlocks,
  fixUnorderedBlocks,
  fixChainSpec,
  fixDescendantValues
);

export { fixDescendantValues, fixPrematureBlocks, fixUnorderedBlocks, fixUnorderedEvents, parsed, patchChainHeadEvents, translate, unpinHash, withNumericIds, withPolkadotSdkCompat };
