// @polkadot/util/cjs/array/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arrayZip = exports.arrayUnzip = exports.arrayShuffle = exports.arrayRange = exports.arrayFlatten = exports.arrayFilter = exports.arrayChunk = void 0;
/**
 * @summary Utility methods that operates on arrays
 */
var chunk_js_1 = require("./chunk.js");
Object.defineProperty(exports, "arrayChunk", { enumerable: true, get: function () { return chunk_js_1.arrayChunk; } });
var filter_js_1 = require("./filter.js");
Object.defineProperty(exports, "arrayFilter", { enumerable: true, get: function () { return filter_js_1.arrayFilter; } });
var flatten_js_1 = require("./flatten.js");
Object.defineProperty(exports, "arrayFlatten", { enumerable: true, get: function () { return flatten_js_1.arrayFlatten; } });
var range_js_1 = require("./range.js");
Object.defineProperty(exports, "arrayRange", { enumerable: true, get: function () { return range_js_1.arrayRange; } });
var shuffle_js_1 = require("./shuffle.js");
Object.defineProperty(exports, "arrayShuffle", { enumerable: true, get: function () { return shuffle_js_1.arrayShuffle; } });
var unzip_js_1 = require("./unzip.js");
Object.defineProperty(exports, "arrayUnzip", { enumerable: true, get: function () { return unzip_js_1.arrayUnzip; } });
var zip_js_1 = require("./zip.js");
Object.defineProperty(exports, "arrayZip", { enumerable: true, get: function () { return zip_js_1.arrayZip; } });
