// @polkadot-api/sdk-ink/dist/esm/src/ink-sdk.mjs
import { Enum, Binary } from 'polkadot-api';
import { wndAh as wndAh_default } from '../.papi/descriptors/dist/index.mjs';
import { inkEncoding, solEncoding } from './encoding-provider.mjs';
import { getContract } from './get-contract.mjs';
import { getDeployer } from './get-deployer.mjs';
import { reviveProvider } from './provider.mjs';
import { getAccountId } from './revive-sdk.mjs';
import { defaultOptions } from './sdk-types.mjs';
import { reviveAddressIsMapped } from './util.mjs';

const createInkSdk = (client, options) => {
  const typedApi = client.getTypedApi(wndAh_default);
  const { atBest } = { ...defaultOptions, ...options };
  const provider = reviveProvider(typedApi, atBest);
  const getContractSdk = (encodingProvider, address) => {
    return getContract(
      provider,
      encodingProvider,
      Binary.fromHex(address),
      (v) => v.asHex(),
      getAccountId(address)
    );
  };
  const getDeployerSdk = (contractDescriptors, code) => {
    const encodingProvider = contractDescriptors.metadata ? inkEncoding(contractDescriptors) : solEncoding(contractDescriptors);
    return getDeployer(
      provider,
      encodingProvider,
      Enum("Upload", code),
      (v) => v.asHex()
    );
  };
  const curriedGetContract = ((contractDescriptors, address) => {
    const encodingProvider = contractDescriptors.metadata ? inkEncoding(contractDescriptors) : solEncoding(contractDescriptors);
    if (address == null) {
      return (address2) => getContractSdk(encodingProvider, address2);
    }
    return getContractSdk(encodingProvider, address);
  });
  return {
    addressIsMapped(address) {
      return reviveAddressIsMapped(typedApi, address);
    },
    getContract: curriedGetContract,
    getDeployer: getDeployerSdk,
    readDeploymentEvents: (contractDescriptors, events) => {
      const encodingProvider = contractDescriptors.metadata ? inkEncoding(contractDescriptors) : solEncoding(contractDescriptors);
      const instantiatedEvents = events?.filter(
        (evt) => evt.type === "Revive" && evt.value.type === "Instantiated"
      ).map(
        (v) => v.value.value
      ) ?? [];
      return instantiatedEvents.map((evt) => ({
        address: evt.contract.asHex(),
        contractEvents: encodingProvider.filterEvents(
          evt.contract.asHex(),
          events
        )
      }));
    }
  };
};

export { createInkSdk };
