// @polkadot-api/sdk-ink/dist/esm/src/inkv5-sdk.mjs
import { Enum } from 'polkadot-api';
import { inkEncoding } from './encoding-provider.mjs';
import { getContract } from './get-contract.mjs';
import { getDeployer } from './get-deployer.mjs';
import { contractsProvider } from './provider.mjs';
import { defaultOptions } from './sdk-types.mjs';

const createInkV5Sdk = (typedApi, contractDescriptors, options) => {
  const { atBest } = { ...defaultOptions, ...options };
  const provider = contractsProvider(typedApi, atBest);
  const encodingProvider = inkEncoding(contractDescriptors);
  return {
    getContract: (address) => getContract(provider, encodingProvider, address, (v) => v, address),
    getDeployer: (code) => getDeployer(provider, encodingProvider, Enum("Upload", code), (v) => v),
    readDeploymentEvents(events) {
      const instantiatedEvents = events?.filter(
        (evt) => evt.type === "Contracts" && evt.value.type === "Instantiated"
      ).map(
        (v) => v.value.value
      ) ?? [];
      return instantiatedEvents.map((evt) => ({
        address: evt.contract,
        contractEvents: encodingProvider.filterEvents(evt.contract, events)
      }));
    }
  };
};

export { createInkV5Sdk };
