// @polkadot-api/sdk-ink/dist/esm/src/revive-sdk.mjs
import { Enum, Binary, AccountId } from 'polkadot-api';
import { mergeUint8 } from 'polkadot-api/utils';
import { inkEncoding, solEncoding } from './encoding-provider.mjs';
import { getContract } from './get-contract.mjs';
import { getDeployer } from './get-deployer.mjs';
import { reviveProvider } from './provider.mjs';
import { defaultOptions } from './sdk-types.mjs';
import { ss58ToEthereum } from './util.mjs';

const createReviveSdk = (typedApi, contractDescriptors, options) => {
  const { atBest } = { ...defaultOptions, ...options };
  const provider = reviveProvider(typedApi, atBest);
  const encodingProvider = contractDescriptors.metadata ? inkEncoding(contractDescriptors) : solEncoding(contractDescriptors);
  return {
    getContract: (address) => getContract(
      provider,
      encodingProvider,
      Binary.fromHex(address),
      (v) => v.asHex(),
      getAccountId(address)
    ),
    getDeployer: (code) => getDeployer(
      provider,
      encodingProvider,
      Enum("Upload", code),
      (v) => v.asHex()
    ),
    readDeploymentEvents(events) {
      const contractEmittedEvents = events?.filter(
        (evt) => evt.type === "Revive" && evt.value.type === "ContractEmitted"
      ) ?? [];
      const contractAddresses = Array.from(
        new Set(
          contractEmittedEvents.map((evt) => evt.value.value?.contract?.asHex?.()).filter((v) => !!v)
        )
      );
      return contractAddresses.map((address) => ({
        address,
        contractEvents: encodingProvider.filterEvents(
          address,
          contractEmittedEvents
        )
      }));
    },
    addressIsMapped: (address) => typedApi.query.Revive.OriginalAccount.getValue(
      ss58ToEthereum(address),
      atBest ? { at: "best" } : {}
    ).then((r) => r != null)
  };
};
const getAccountId = (address) => {
  const publicKey = mergeUint8([
    Binary.fromHex(address).asBytes(),
    new Uint8Array(new Array(32 - 20).fill(238))
  ]);
  return AccountId().dec(publicKey);
};

export { createReviveSdk, getAccountId };
