// viem/_cjs/zksync/constants/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addressModulo = exports.bootloaderFormalAddress = exports.l2NativeTokenVaultAddress = exports.l2AssetRouterAddress = exports.l1ToL2AliasOffset = exports.l1MessengerAddress = exports.l2BaseTokenAddress = exports.ethAddressInContracts = exports.legacyEthAddress = exports.contract2FactoryAddress = exports.contractDeployerAddress = void 0;
exports.contractDeployerAddress = '0x0000000000000000000000000000000000008006';
exports.contract2FactoryAddress = '0x0000000000000000000000000000000000010000';
exports.legacyEthAddress = '0x0000000000000000000000000000000000000000';
exports.ethAddressInContracts = '0x0000000000000000000000000000000000000001';
exports.l2BaseTokenAddress = '0x000000000000000000000000000000000000800a';
exports.l1MessengerAddress = '0x0000000000000000000000000000000000008008';
exports.l1ToL2AliasOffset = '0x1111000000000000000000000000000000001111';
exports.l2AssetRouterAddress = '0x0000000000000000000000000000000000010003';
exports.l2NativeTokenVaultAddress = '0x0000000000000000000000000000000000010004';
exports.bootloaderFormalAddress = '0x0000000000000000000000000000000000008001';
exports.addressModulo = 2n ** 160n;
