// @polkadot/types-codec/cjs/primitive/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.usize = exports.USize = exports.u256 = exports.U256 = exports.u128 = exports.U128 = exports.u64 = exports.U64 = exports.u32 = exports.U32 = exports.u16 = exports.U16 = exports.u8 = exports.U8 = exports.isize = exports.ISize = exports.i256 = exports.I256 = exports.i128 = exports.I128 = exports.i64 = exports.I64 = exports.i32 = exports.I32 = exports.i16 = exports.I16 = exports.i8 = exports.I8 = exports.f64 = exports.F64 = exports.f32 = exports.F32 = void 0;
var F32_js_1 = require("./F32.js");
Object.defineProperty(exports, "F32", { enumerable: true, get: function () { return F32_js_1.f32; } });
Object.defineProperty(exports, "f32", { enumerable: true, get: function () { return F32_js_1.f32; } });
var F64_js_1 = require("./F64.js");
Object.defineProperty(exports, "F64", { enumerable: true, get: function () { return F64_js_1.f64; } });
Object.defineProperty(exports, "f64", { enumerable: true, get: function () { return F64_js_1.f64; } });
var I8_js_1 = require("./I8.js");
Object.defineProperty(exports, "I8", { enumerable: true, get: function () { return I8_js_1.i8; } });
Object.defineProperty(exports, "i8", { enumerable: true, get: function () { return I8_js_1.i8; } });
var I16_js_1 = require("./I16.js");
Object.defineProperty(exports, "I16", { enumerable: true, get: function () { return I16_js_1.i16; } });
Object.defineProperty(exports, "i16", { enumerable: true, get: function () { return I16_js_1.i16; } });
var I32_js_1 = require("./I32.js");
Object.defineProperty(exports, "I32", { enumerable: true, get: function () { return I32_js_1.i32; } });
Object.defineProperty(exports, "i32", { enumerable: true, get: function () { return I32_js_1.i32; } });
var I64_js_1 = require("./I64.js");
Object.defineProperty(exports, "I64", { enumerable: true, get: function () { return I64_js_1.i64; } });
Object.defineProperty(exports, "i64", { enumerable: true, get: function () { return I64_js_1.i64; } });
var I128_js_1 = require("./I128.js");
Object.defineProperty(exports, "I128", { enumerable: true, get: function () { return I128_js_1.i128; } });
Object.defineProperty(exports, "i128", { enumerable: true, get: function () { return I128_js_1.i128; } });
var I256_js_1 = require("./I256.js");
Object.defineProperty(exports, "I256", { enumerable: true, get: function () { return I256_js_1.i256; } });
Object.defineProperty(exports, "i256", { enumerable: true, get: function () { return I256_js_1.i256; } });
var ISize_js_1 = require("./ISize.js");
Object.defineProperty(exports, "ISize", { enumerable: true, get: function () { return ISize_js_1.isize; } });
Object.defineProperty(exports, "isize", { enumerable: true, get: function () { return ISize_js_1.isize; } });
var U8_js_1 = require("./U8.js");
Object.defineProperty(exports, "U8", { enumerable: true, get: function () { return U8_js_1.u8; } });
Object.defineProperty(exports, "u8", { enumerable: true, get: function () { return U8_js_1.u8; } });
var U16_js_1 = require("./U16.js");
Object.defineProperty(exports, "U16", { enumerable: true, get: function () { return U16_js_1.u16; } });
Object.defineProperty(exports, "u16", { enumerable: true, get: function () { return U16_js_1.u16; } });
var U32_js_1 = require("./U32.js");
Object.defineProperty(exports, "U32", { enumerable: true, get: function () { return U32_js_1.u32; } });
Object.defineProperty(exports, "u32", { enumerable: true, get: function () { return U32_js_1.u32; } });
var U64_js_1 = require("./U64.js");
Object.defineProperty(exports, "U64", { enumerable: true, get: function () { return U64_js_1.u64; } });
Object.defineProperty(exports, "u64", { enumerable: true, get: function () { return U64_js_1.u64; } });
var U128_js_1 = require("./U128.js");
Object.defineProperty(exports, "U128", { enumerable: true, get: function () { return U128_js_1.u128; } });
Object.defineProperty(exports, "u128", { enumerable: true, get: function () { return U128_js_1.u128; } });
var U256_js_1 = require("./U256.js");
Object.defineProperty(exports, "U256", { enumerable: true, get: function () { return U256_js_1.u256; } });
Object.defineProperty(exports, "u256", { enumerable: true, get: function () { return U256_js_1.u256; } });
var USize_js_1 = require("./USize.js");
Object.defineProperty(exports, "USize", { enumerable: true, get: function () { return USize_js_1.usize; } });
Object.defineProperty(exports, "usize", { enumerable: true, get: function () { return USize_js_1.usize; } });
