// tsc-prog/dist/utils/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Color = exports.logDiagnostics = void 0;
const ts = require("typescript");
/**
 * Log compiler diagnostics to stderr.
 * @internal
 */
function logDiagnostics(diagnostics, pretty = false) {
    if (!diagnostics.length)
        return;
    const formatHost = {
        getCanonicalFileName: (path) => path,
        getCurrentDirectory: ts.sys.getCurrentDirectory,
        getNewLine: () => ts.sys.newLine,
    };
    const message = pretty
        ? ts.formatDiagnosticsWithColorAndContext(diagnostics, formatHost)
        : ts.formatDiagnostics(diagnostics, formatHost);
    console.warn(message);
}
exports.logDiagnostics = logDiagnostics;
/**
 * @internal
 */
var Color;
(function (Color) {
    const isTTY = !!ts.sys.writeOutputIsTTY && ts.sys.writeOutputIsTTY();
    Color.grey = (text) => (isTTY ? EscSeq.Grey + text + EscSeq.Reset : text);
    Color.red = (text) => (isTTY ? EscSeq.Red + text + EscSeq.Reset : text);
    Color.green = (text) => (isTTY ? EscSeq.Green + text + EscSeq.Reset : text);
    Color.yellow = (text) => (isTTY ? EscSeq.Yellow + text + EscSeq.Reset : text);
    Color.blue = (text) => (isTTY ? EscSeq.Blue + text + EscSeq.Reset : text);
    Color.magenta = (text) => (isTTY ? EscSeq.Magenta + text + EscSeq.Reset : text);
    Color.cyan = (text) => (isTTY ? EscSeq.Cyan + text + EscSeq.Reset : text);
    /**
     * Foreground color escape sequences.
     * Taken from https://github.com/microsoft/TypeScript/blob/v3.5.3/src/compiler/program.ts#L368
     */
    let EscSeq;
    (function (EscSeq) {
        EscSeq["Grey"] = "\u001B[90m";
        EscSeq["Red"] = "\u001B[91m";
        EscSeq["Green"] = "\u001B[92m";
        EscSeq["Yellow"] = "\u001B[93m";
        EscSeq["Blue"] = "\u001B[94m";
        EscSeq["Magenta"] = "\u001B[95m";
        EscSeq["Cyan"] = "\u001B[96m";
        EscSeq["Reset"] = "\u001B[0m";
    })(EscSeq || (EscSeq = {}));
})(Color = exports.Color || (exports.Color = {}));
