// @polkadot-api/known-chains/dist/index.mjs
import {
  chainSpec as chainSpec24
} from "./chunk-OEEEU7AE.mjs";
import {
  chainSpec as chainSpec25
} from "./chunk-4SLS4OOK.mjs";
import {
  chainSpec as chainSpec26
} from "./chunk-2XKAWMHH.mjs";
import {
  chainSpec as chainSpec16
} from "./chunk-V5VATG3Q.mjs";
import {
  chainSpec as chainSpec17
} from "./chunk-3NBF63QS.mjs";
import {
  chainSpec as chainSpec18
} from "./chunk-QIV5AJQF.mjs";
import {
  chainSpec as chainSpec19
} from "./chunk-TP7HBWY6.mjs";
import {
  chainSpec as chainSpec20
} from "./chunk-URFAVCF4.mjs";
import {
  chainSpec as chainSpec21
} from "./chunk-VRLIQPBU.mjs";
import {
  chainSpec as chainSpec22
} from "./chunk-BXH7RC66.mjs";
import {
  chainSpec as chainSpec23
} from "./chunk-HD3R7KJJ.mjs";
import {
  chainSpec as chainSpec8
} from "./chunk-DPYHF5EQ.mjs";
import {
  chainSpec as chainSpec9
} from "./chunk-GTJJX2M2.mjs";
import {
  chainSpec as chainSpec10
} from "./chunk-5VZIQBMS.mjs";
import {
  chainSpec as chainSpec11
} from "./chunk-D2WQ7WJ7.mjs";
import {
  chainSpec as chainSpec12
} from "./chunk-JXLVOOH5.mjs";
import {
  chainSpec as chainSpec13
} from "./chunk-SOINKWG7.mjs";
import {
  chainSpec as chainSpec14
} from "./chunk-AVS4PDRR.mjs";
import {
  chainSpec as chainSpec15
} from "./chunk-WPYCR7LE.mjs";
import {
  chainSpec
} from "./chunk-FLYNJO2O.mjs";
import {
  chainSpec as chainSpec2
} from "./chunk-DQYTBFVY.mjs";
import {
  chainSpec as chainSpec3
} from "./chunk-QE62BN6Z.mjs";
import {
  chainSpec as chainSpec4
} from "./chunk-C5PIRAJD.mjs";
import {
  chainSpec as chainSpec5
} from "./chunk-D7XKE5XB.mjs";
import {
  chainSpec as chainSpec6
} from "./chunk-4Q5KAEDT.mjs";
import {
  chainSpec as chainSpec7
} from "./chunk-QW4A72JR.mjs";
export {
  chainSpec as ksmcc3,
  chainSpec2 as ksmcc3_asset_hub,
  chainSpec3 as ksmcc3_bridge_hub,
  chainSpec4 as ksmcc3_coretime,
  chainSpec5 as ksmcc3_encointer,
  chainSpec6 as ksmcc3_people,
  chainSpec7 as paseo,
  chainSpec8 as paseo_asset_hub,
  chainSpec9 as paseo_coretime,
  chainSpec10 as paseo_people,
  chainSpec11 as polkadot,
  chainSpec12 as polkadot_asset_hub,
  chainSpec13 as polkadot_bridge_hub,
  chainSpec14 as polkadot_collectives,
  chainSpec15 as polkadot_coretime,
  chainSpec16 as polkadot_people,
  chainSpec17 as rococo_v2_2,
  chainSpec18 as rococo_v2_2_asset_hub,
  chainSpec19 as rococo_v2_2_bridge_hub,
  chainSpec20 as rococo_v2_2_people,
  chainSpec21 as westend2,
  chainSpec22 as westend2_asset_hub,
  chainSpec23 as westend2_bridge_hub,
  chainSpec24 as westend2_collectives,
  chainSpec25 as westend2_coretime,
  chainSpec26 as westend2_people
};
