// @polkadot/util-crypto/cjs/ed25519/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ed25519Verify = exports.ed25519Sign = exports.ed25519PairFromString = exports.ed25519PairFromSeed = exports.ed25519PairFromSecret = exports.ed25519PairFromRandom = exports.ed25519DeriveHard = void 0;
/**
 * @summary Implements ed25519 operations
 */
var deriveHard_js_1 = require("./deriveHard.js");
Object.defineProperty(exports, "ed25519DeriveHard", { enumerable: true, get: function () { return deriveHard_js_1.ed25519DeriveHard; } });
var fromRandom_js_1 = require("./pair/fromRandom.js");
Object.defineProperty(exports, "ed25519PairFromRandom", { enumerable: true, get: function () { return fromRandom_js_1.ed25519PairFromRandom; } });
var fromSecret_js_1 = require("./pair/fromSecret.js");
Object.defineProperty(exports, "ed25519PairFromSecret", { enumerable: true, get: function () { return fromSecret_js_1.ed25519PairFromSecret; } });
var fromSeed_js_1 = require("./pair/fromSeed.js");
Object.defineProperty(exports, "ed25519PairFromSeed", { enumerable: true, get: function () { return fromSeed_js_1.ed25519PairFromSeed; } });
var fromString_js_1 = require("./pair/fromString.js");
Object.defineProperty(exports, "ed25519PairFromString", { enumerable: true, get: function () { return fromString_js_1.ed25519PairFromString; } });
var sign_js_1 = require("./sign.js");
Object.defineProperty(exports, "ed25519Sign", { enumerable: true, get: function () { return sign_js_1.ed25519Sign; } });
var verify_js_1 = require("./verify.js");
Object.defineProperty(exports, "ed25519Verify", { enumerable: true, get: function () { return verify_js_1.ed25519Verify; } });
